---
title: "R_assignment_README"
author: "Molly Bledsoe"
date: "3/18/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Data Inspection: fang_et_al_genotypes.txt

##Downloading the fang_et_al_genotypes.txt file from the repository:
```{r}
fang_genotypes <- read.csv("https://raw.githubusercontent.com/EEOB-BioData/BCB546-Spring2021/master/assignments/UNIX_Assignment/fang_et_al_genotypes.txt", header=T, sep="\t", stringsAsFactors =F)
```

##Commands used to inspect fang_genotypes:
```{r}
  typeof(fang_genotypes)
  file.size("fang_et_al_genotypes.txt")
  length(fang_genotypes)
  dim(fang_genotypes)
  nrow(fang_genotypes)
  ncol(fang_genotypes)
  names(fang_genotypes)
  class(fang_genotypes)
  sapply(fang_genotypes,class)
  str(fang_genotypes)
```

##What I learned from inspecting this file:
##This file is a list
##This file has a size of 11054722 
##This file contains 986 elements
##This file has dimensions of 2782 x 986
##This file has 2782 rows
##This file has 986 columns
##This file's class is data.frame
##his file contains base-pairs (known and unknown)found on chromosomes and the Sample_ID
   
#Data Inspection: snp_position.txt

##Downloading the snp_position.txt file from the repository:

```{r}
snp_position <- read.csv("https://raw.githubusercontent.com/EEOB-BioData/BCB546-Spring2021/master/assignments/UNIX_Assignment/snp_position.txt", header=T, sep="\t", stringsAsFactors =F)
```

##Commands used to inspect snp:
```{r}
  typeof(snp_position)
  file.size("snp_position.txt")
  length(snp_position)
  dim(snp_position)
  nrow(snp_position)
  ncol(snp_position)
  names(snp_position)
  class(snp_position)
  sapply(snp_position,class)
  str(snp_position)
```  

##What I learned from inspecting this file:
##This file is a list
##This file has a size of 83747
##This file contains 15 elements
##This file has dimensions of 983 x 15
##This file has 983 rows
##This file has 15 columns
##The first row contains SNP_ID
##The third row contains the Chromosome numbers
##The forth row contains the positions

    
# Data Processing:

##Grab the maize and teosinte groups:
```{r}
library(dplyr)
maize <- fang_genotypes[fang_genotypes$Group=="ZMMIL" | fang_genotypes$Group=="ZMMLR" | fang_genotypes$Group=="ZMMMR",]
  
teosinte <- fang_genotypes[fang_genotypes$Group=="ZMPBA" | fang_genotypes$Group=="ZMPIL" | fang_genotypes$Group=="ZMPJA",]  
``` 
 
##Grab the desired snp_position columns:
```{r}
snp_organized <- snp_position[c(1,3,4)]
```
##Transpose maize and teosinte files:
```{r}
library(tidyverse)
maize_transposed <- as.data.frame(t(maize))
colnames(maize_transposed) <- maize$Sample_ID

teosinte_transposed <- as.data.frame(t(maize))
colnames(maize_transposed) <- maize$Sample_ID
```

##Join the maize/teosinte_transposed files with the snp_organized file:

```{r}
joined_snp_maize_transposed <- merge(x = snp_organized,y = maize_transposed, by.x="SNP_ID", by.y="row.names", all.y = TRUE)

joined_snp_teosinte_transposed <- merge(x = snp_organized,y = teosinte_transposed, by.x="SNP_ID", by.y="row.names", all.y = TRUE)
```
##Set Chromosome and Position values as numeric:
```{r}
joined_snp_maize_transposed$Position <- as.numeric(joined_snp_maize_transposed$Position)
joined_snp_maize_transposed$Chromosome <- as.numeric(joined_snp_maize_transposed$Chromosome)

joined_snp_teosinte_transposed$Position <- as.numeric(joined_snp_teosinte_transposed$Position)
joined_snp_teosinte_transposed$Chromosome <- as.numeric(joined_snp_teosinte_transposed$Chromosome)
```

##Remove SNPS at "unknown" and "multiple" Positions and Chromosomes:
```{r}
joined_maize <- joined_snp_maize_transposed[!(joined_snp_maize_transposed$Position == "unknown") | !(joined_snp_maize_transposed$Chromosome == "unknown")| !(joined_snp_maize_transposed$Chromosome == "multiple"),]
joined_maize <- na.omit(joined_maize)

joined_teosinte <- joined_snp_teosinte_transposed[!(joined_snp_teosinte_transposed$Position == "unknown") | !(joined_snp_teosinte_transposed$Chromosome == "unknown")| !(joined_snp_teosinte_transposed$Chromosome == "multiple"),]
joined_teosinte <- na.omit(joined_maize)
```

##Filter joined files by Chromosome numbers, put in ascending position order, and organize into the proper folders:
```{r}
asc_maize_chr1 <- subset(joined_maize, Chromosome ==1)%>%arrange(Position)%>%write_csv(file= "Maize_files/asc_maize_chr1.txt")
asc_maize_chr2 <- subset(joined_maize, Chromosome ==2)%>%arrange(Position)%>%write_csv(file= "Maize_files/asc_maize_chr2.txt")
asc_maize_chr3 <- subset(joined_maize, Chromosome ==3)%>%arrange(Position)%>%write_csv(file= "Maize_files/asc_maize_chr3.txt")
asc_maize_chr4 <- subset(joined_maize, Chromosome ==4)%>%arrange(Position)%>%write_csv(file= "Maize_files/asc_maize_chr4.txt")
asc_maize_chr5 <- subset(joined_maize, Chromosome ==5)%>%arrange(Position)%>%write_csv(file= "Maize_files/asc_maize_chr5.txt")
asc_maize_chr6 <- subset(joined_maize, Chromosome ==6)%>%arrange(Position)%>%write_csv(file= "Maize_files/asc_maize_chr6.txt")
asc_maize_chr7 <- subset(joined_maize, Chromosome ==7)%>%arrange(Position)%>%write_csv(file= "Maize_files/asc_maize_chr7.txt")
asc_maize_chr8 <- subset(joined_maize, Chromosome ==8)%>%arrange(Position)%>%write_csv(file= "Maize_files/asc_maize_chr8.txt")
asc_maize_chr9 <- subset(joined_maize, Chromosome ==9)%>%arrange(Position)%>%write_csv(file= "Maize_files/asc_maize_chr9.txt")
asc_maize_chr10 <- subset(joined_maize, Chromosome ==10)%>%arrange(Position)%>%arrange(Position)%>%write_csv(file= "Maize_files/asc_maize_chr10.txt")

asc_teosinte_chr1 <- subset(joined_teosinte, Chromosome ==1)%>%arrange(Position)%>%arrange(Position)%>%write_csv(file= "Teosinte_files/asc_teosinte_chr1.txt")
asc_teosinte_chr2 <- subset(joined_teosinte, Chromosome ==2)%>%arrange(Position)%>%arrange(Position)%>%write_csv(file= "Teosinte_files/asc_teosinte_chr2.txt")
asc_teosinte_chr3 <- subset(joined_teosinte, Chromosome ==3)%>%arrange(Position)%>%arrange(Position)%>%write_csv(file= "Teosinte_files/asc_teosinte_chr3.txt")
asc_teosinte_chr4 <- subset(joined_teosinte, Chromosome ==4)%>%arrange(Position)%>%arrange(Position)%>%write_csv(file= "Teosinte_files/asc_teosinte_chr4.txt")
asc_teosinte_chr5 <- subset(joined_teosinte, Chromosome ==5)%>%arrange(Position)%>%arrange(Position)%>%write_csv(file= "Teosinte_files/asc_teosinte_chr5.txt")
asc_teosinte_chr6 <- subset(joined_teosinte, Chromosome ==6)%>%arrange(Position)%>%arrange(Position)%>%write_csv(file= "Teosinte_files/asc_teosinte_chr6.txt")
asc_teosinte_chr7 <- subset(joined_teosinte, Chromosome ==7)%>%arrange(Position)%>%arrange(Position)%>%write_csv(file= "Teosinte_files/asc_teosinte_chr7.txt")
asc_teosinte_chr8 <- subset(joined_teosinte, Chromosome ==8)%>%arrange(Position)%>%arrange(Position)%>%write_csv(file= "Teosinte_files/asc_teosinte_chr8.txt")
asc_teosinte_chr9 <- subset(joined_teosinte, Chromosome ==9)%>%arrange(Position)%>%arrange(Position)%>%write_csv(file= "Teosinte_files/asc_teosinte_chr9.txt")
asc_teosinte_chr10 <- subset(joined_teosinte, Chromosome ==10)%>%arrange(Position)%>%arrange(Position)%>%write_csv(file= "Teosinte_files/asc_teosinte_chr10.txt")
```

##Filter joined files by Chromosome numbers and put in descending position order:
```{r}
desc_maize_chr1 <- subset(joined_maize, Chromosome ==1)%>%arrange(desc(Position))
desc_maize_chr2 <- subset(joined_maize, Chromosome ==2)%>%arrange(desc(Position))
desc_maize_chr3 <- subset(joined_maize, Chromosome ==3)%>%arrange(desc(Position))
desc_maize_chr4 <- subset(joined_maize, Chromosome ==4)%>%arrange(desc(Position))
desc_maize_chr5 <- subset(joined_maize, Chromosome ==5)%>%arrange(desc(Position))
desc_maize_chr6 <- subset(joined_maize, Chromosome ==6)%>%arrange(desc(Position))
desc_maize_chr7 <- subset(joined_maize, Chromosome ==7)%>%arrange(desc(Position))
desc_maize_chr8 <- subset(joined_maize, Chromosome ==8)%>%arrange(desc(Position))
desc_maize_chr9 <- subset(joined_maize, Chromosome ==9)%>%arrange(desc(Position))
desc_maize_chr10 <- subset(joined_maize, Chromosome ==10)%>%arrange(desc(Position))

desc_teosinte_chr1 <- subset(joined_teosinte, Chromosome ==1)%>%arrange(desc(Position))
desc_teosinte_chr2 <- subset(joined_teosinte, Chromosome ==2)%>%arrange(desc(Position))
desc_teosinte_chr3 <- subset(joined_teosinte, Chromosome ==3)%>%arrange(desc(Position))
desc_teosinte_chr4 <- subset(joined_teosinte, Chromosome ==4)%>%arrange(desc(Position))
desc_teosinte_chr5 <- subset(joined_teosinte, Chromosome ==5)%>%arrange(desc(Position))
desc_teosinte_chr6 <- subset(joined_teosinte, Chromosome ==6)%>%arrange(desc(Position))
desc_teosinte_chr7 <- subset(joined_teosinte, Chromosome ==7)%>%arrange(desc(Position))
desc_teosinte_chr8 <- subset(joined_teosinte, Chromosome ==8)%>%arrange(desc(Position))
desc_teosinte_chr9 <- subset(joined_teosinte, Chromosome ==9)%>%arrange(desc(Position))
desc_teosinte_chr10 <- subset(joined_teosinte, Chromosome ==10)%>%arrange(desc(Position))
```

##Changing missing values from "?" to "-" in the decreasing files:

```{r}
desc_maize_chr1[desc_maize_chr1 == '?/?'] <- '-/-'
desc_maize_chr2[desc_maize_chr2 == '?/?'] <- '-/-'
desc_maize_chr3[desc_maize_chr3 == '?/?'] <- '-/-'
desc_maize_chr4[desc_maize_chr4 == '?/?'] <- '-/-'
desc_maize_chr5[desc_maize_chr5 == '?/?'] <- '-/-'
desc_maize_chr6[desc_maize_chr6 == '?/?'] <- '-/-'
desc_maize_chr7[desc_maize_chr7 == '?/?'] <- '-/-'
desc_maize_chr8[desc_maize_chr8 == '?/?'] <- '-/-'
desc_maize_chr9[desc_maize_chr9 == '?/?'] <- '-/-'
desc_maize_chr10[desc_maize_chr10 == '?/?'] <- '-/-'

desc_teosinte_chr1[desc_teosinte_chr1 == '?/?'] <- '-/-'
desc_teosinte_chr2[desc_teosinte_chr2 == '?/?'] <- '-/-'
desc_teosinte_chr3[desc_teosinte_chr3 == '?/?'] <- '-/-'
desc_teosinte_chr4[desc_teosinte_chr4 == '?/?'] <- '-/-'
desc_teosinte_chr5[desc_teosinte_chr5 == '?/?'] <- '-/-'
desc_teosinte_chr6[desc_teosinte_chr6 == '?/?'] <- '-/-'
desc_teosinte_chr7[desc_teosinte_chr7 == '?/?'] <- '-/-'
desc_teosinte_chr8[desc_teosinte_chr8 == '?/?'] <- '-/-'
desc_teosinte_chr9[desc_teosinte_chr9 == '?/?'] <- '-/-'
desc_teosinte_chr10[desc_teosinte_chr10 == '?/?'] <- '-/-'
```

#Move the descending data into the proper folders:
```{r}
desc_maize_chr1%>%write_csv(file= "Maize_files/desc_maize_chr1.txt")
desc_maize_chr2%>%write_csv(file= "Maize_files/desc_maize_chr2.txt")
desc_maize_chr3%>%write_csv(file= "Maize_files/desc_maize_chr3.txt")
desc_maize_chr4%>%write_csv(file= "Maize_files/desc_maize_chr4.txt")
desc_maize_chr5%>%write_csv(file= "Maize_files/desc_maize_chr5.txt")
desc_maize_chr6%>%write_csv(file= "Maize_files/desc_maize_chr6.txt")
desc_maize_chr7%>%write_csv(file= "Maize_files/desc_maize_chr7.txt")
desc_maize_chr8%>%write_csv(file= "Maize_files/desc_maize_chr8.txt")
desc_maize_chr9%>%write_csv(file= "Maize_files/desc_maize_chr9.txt")
desc_maize_chr10%>%write_csv(file= "Maize_files/desc_maize_chr10.txt")

desc_teosinte_chr1%>%write_csv(file= "Teosinte_files/desc_teosinte_chr1.txt")
desc_teosinte_chr2%>%write_csv(file= "Teosinte_files/desc_teosinte_chr2.txt")
desc_teosinte_chr3%>%write_csv(file= "Teosinte_files/desc_teosinte_chr3.txt")
desc_teosinte_chr4%>%write_csv(file= "Teosinte_files/desc_teosinte_chr4.txt")
desc_teosinte_chr5%>%write_csv(file= "Teosinte_files/desc_teosinte_chr5.txt")
desc_teosinte_chr6%>%write_csv(file= "Teosinte_files/desc_teosinte_chr6.txt")
desc_teosinte_chr7%>%write_csv(file= "Teosinte_files/desc_teosinte_chr7.txt")
desc_teosinte_chr8%>%write_csv(file= "Teosinte_files/desc_teosinte_chr8.txt")
desc_teosinte_chr9%>%write_csv(file= "Teosinte_files/desc_teosinte_chr9.txt")
desc_teosinte_chr10%>%write_csv(file= "Teosinte_files/desc_teosinte_chr10.txt")
```
#Part Two: Plotting

##Transpose our full dataset:
```{r}
fang_transposed <- as.data.frame(t(fang_genotypes))
colnames(fang_transposed) <- fang_genotypes$Sample_ID
```
##Join the tranposed file with the organizes SNP file:
```{r}
joined_fang <- merge(x = snp_organized,y = fang_transposed, by.x="SNP_ID", by.y="row.names", all.y = TRUE)
```
##Plot One - SNPs per Chromosome:
```{r}
joined_fang$Chromosome <- factor(joined_fang$Chromosome, levels = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "multiple", "unknown", "NA"))

snps_per_chromosome_plot <- ggplot(joined_fang, aes(x= Chromosome) ) + geom_histogram(stat= "Count", color = "blue", fill = "white") + ggtitle ("Number of SNPs per Chromosome")

snps_per_chromosome_plot

ggsave(filename = "Graphs/snps_per_chromosome.png", device="png")
```
##Plot Two - SNP Distribution on each Chromosome:
```{r}
snps_distribution_per_chromosome_plot <- ggplot(joined_fang, aes(x= Chromosome, y= Position))+ geom_point(stat=, color = "red", alpha= 0.01)+ ggtitle ("SNPs Distribution per Chromosome")+theme(axis.text.y=element_blank(),axis.ticks.y=element_blank())

snps_distribution_per_chromosome_plot

ggsave(filename = "Graphs/snps_distribution_per_chromosome.png", device="png")
```
##Plot Three - Amount of Heterozygosity:

###Melt the data set:
```{r}
library(reshape2)
fang_headers <- colnames(fang_genotypes)[-c(1:3)]
melted_fang <- melt(fang_genotypes, measure.vars = fang_headers)
```
###Rename missing values:
```{r}
melted_fang[ melted_fang == "?/?"]= "No_Data"
```
###Create new column for Homozygous values:
```{r}
melted_fang$Homozygous <- (melted_fang$value=="A/A" | melted_fang$value== "C/C" | melted_fang$value== "G/G" | melted_fang$value== "T/T")
```
###Organize the dataframe by Sample IDs:
```{r}
fang_samples_organized <- melted_fang[order(melted_fang$Sample_ID),]

library(plyr)

amounts_homozygous_heterozygous <- ddply(fang_samples_organized, c("Sample_ID"), summarise, heterozygous_amount=sum(!Homozygous, na.rm=TRUE), homozygous_amount=sum(Homozygous, na.rm=TRUE), No_Data=sum(is.na(Homozygous)))

melted_amounts <- melt(amounts_homozygous_heterozygous, measure.vars= c("heterozygous_amount", "homozygous_amount", "No_Data"))
```
###Making the Missing Data and Amount of Heterozygosity Plot:
```{r}
missing_heterozygosity_plot <- ggplot(melted_amounts, aes(x= Sample_ID, y= value, fill=variable)) + geom_bar(stat= "identity", position = "stack") + ggtitle("Missing Data and Amount of Heterozygosity")

missing_heterozygosity_plot

ggsave(filename="graphs/missing_heterozygosity.png", device="png")
```
##Plot Four - Missing Data and Amount of Heterozygosity per Group:

###Organize the datafram by groups:
```{r}
fang_groups_organized <- melted_fang[order(melted_fang$Group),]

amounts_groups <- ddply(fang_groups_organized, c("Group"), summarise, heterozygous_amount=sum(!Homozygous, na.rm=TRUE), homozygous_amount=sum(Homozygous, na.rm=TRUE), No_Data=sum(is.na(Homozygous)))

melted_groups <- melt(amounts_groups, measure.vars= c("heterozygous_amount", "homozygous_amount", "No_Data"))
```
###Making the plot:
```{r}
missing_data_in_groups_plot <- ggplot(melted_groups, aes(x= Group, y= value, fill= variable)) + geom_bar(stat="identity", position = "fill") + ggtitle("Missing data and Amount of Heterozygosity per Group")+theme(axis.text.x = element_text(angle = 90))

missing_data_in_groups_plot

ggsave(filename="Graphs/missing_data_in_groups.png", device="png")
```
##Making my own Visualization - Plotting How many samples per Group
```{r}
samples_per_group_plot <- ggplot(melted_fang, aes(x= Group))+geom_bar(stat= "Count", color = "purple", fill = "purple") +ggtitle ("Number of Samples per Group")+ theme(axis.text.x = element_text(angle = 90))

samples_per_group_plot

ggsave(filename="Graphs/samples_per_group.png", device="png")
```

